src/coreclr/jit/valuenum.cpp

```diff
@@ -3651,6 +3668,108 @@ ValueNum ValueNumStore::EvalBitCastForConstantArgs(var_types dstType, ValueNum a
+//    arg1VN - VN of the second argument
+//
+// Return Value:
+//    NoVN if this is not a foldable type compare
+//    Simplified (perhaps constant) VN if it is foldable.
+//
+// Notes:
+//    Value number counterpart to gtFoldTypeCompare
+//    Doesn't handle all the cases (yet).
+//
+//    (EQ/NE (TypeHandleToRuntimeType x) (TypeHandleToRuntimeType y)) == (EQ/NE x y)
+//
+ValueNum ValueNumStore::VNEvalFoldTypeCompare(var_types type, VNFunc func, ValueNum arg0VN, ValueNum arg1VN)
+{
+    const genTreeOps oper = genTreeOps(func);
+    assert(GenTree::StaticOperIs(oper, GT_EQ, GT_NE));
+
+    VNFuncApp  arg0Func;
+    const bool arg0IsFunc = GetVNFunc(arg0VN, &arg0Func);
+
+    if (!arg0IsFunc || (arg0Func.m_func != VNF_TypeHandleToRuntimeType))
+    {
+        return NoVN;
+    }
+
+    VNFuncApp  arg1Func;
+    const bool arg1IsFunc = GetVNFunc(arg1VN, &arg1Func);
+
+    if (!arg1IsFunc || (arg1Func.m_func != VNF_TypeHandleToRuntimeType))
+    {
+        return NoVN;
+    }
+
+    // Only re-express as handle equality when we have known
+    // class handles and the VM agrees comparing these gives the same
+    // result as comparing the runtime types.
+    //
+    // Note that VN actually tracks the value of embedded handle;
+    // we need to pass the VM the associated the compile time handles,
+    // in case they differ (say for prejitting or AOT).
+    //
+    ValueNum handle0 = arg0Func.m_args[0];
+    if (!IsVNHandle(handle0))
+    {
+        return NoVN;
+    }
+
+    ValueNum handle1 = arg1Func.m_args[0];
+    if (!IsVNHandle(handle1))
+    {
+        return NoVN;
+    }
+
+    assert(GetHandleFlags(handle0) == GTF_ICON_CLASS_HDL);
+    assert(GetHandleFlags(handle1) == GTF_ICON_CLASS_HDL);
+
+    const ssize_t handleVal0 = ConstantValue<ssize_t>(handle0);
+    const ssize_t handleVal1 = ConstantValue<ssize_t>(handle1);
+    ssize_t       compileTimeHandle0;
+    ssize_t       compileTimeHandle1;
+
+    // These mappings should always exist.
+    //
+    const bool found0 = m_embeddedToCompileTimeHandleMap.TryGetValue(handleVal0, &compileTimeHandle0);
+    const bool found1 = m_embeddedToCompileTimeHandleMap.TryGetValue(handleVal1, &compileTimeHandle1);
+    assert(found0 && found1);
+
+    // We may see null compile time handles for some constructed class handle cases.
+    // We should fix the construction if possible. But just skip those cases for now.
+    //
+    if ((compileTimeHandle0 == 0) || (compileTimeHandle1 == 0))
+    {
+        return NoVN;
+    }
+
+    JITDUMP("Asking runtime to compare %p (%s) and %p (%s) for equality\n", dspPtr(compileTimeHandle0),
+            m_pComp->eeGetClassName(CORINFO_CLASS_HANDLE(compileTimeHandle0)), dspPtr(compileTimeHandle1),
+            m_pComp->eeGetClassName(CORINFO_CLASS_HANDLE(compileTimeHandle1)));
+
+    ValueNum               result = NoVN;
+    const TypeCompareState s =
+        m_pComp->info.compCompHnd->compareTypesForEquality(CORINFO_CLASS_HANDLE(compileTimeHandle0),
```

Comment:
I'm ok with this, but we should probably open an issue about removing `ICorStaticInfo::canInlineTypeCheck` if we are not going to consult with it here.

Reply:
We don't consult with `canInlineTypeCheck` in the analogous case over in `gtFoldTypeCompare` either.

There is some redundancy with between `compareTyhpesForEquality` and `canInlineTypeCheck`. The former should return `May` for cases where the latter would return something other than `..._PASS`.

Reply:
It is reasonable and probably best to keep this in line with `gtFoldTypeCompare`. But yes, I am pretty confused about the purpose of these JIT-EE functions. Here is what I would expect given their comments/implementations:
* We need to call `canInlineTypeCheck`. It would e.g. be responsible for saying "no" when the types involved do not have an injective mapping for `TYPEHANDLE_TO_RUNTIMETYPE`.
* We do not need to call `compareTypesForEquality`. The only case where it is not `hnd1 == hnd2` is for type handles with shared generics which we cannot see here (they should be runtime lookups). However, it does not hurt to call it and you could imagine it would be necessary for other runtimes or in other cases.

Reply:
Yes, it is a little confusing. There are 4 similar concepts in play
* runtime type
* compile time handle
* embedded handle
* method table

`compareTypesForEquality` is asking about whether one can infer (at compile time) equivalence of runtime types from compile time handles.

`canInlineTypeCheck` is asking whether one can infer (at run time) equivalence of runtime types by comparing an embedded handle to a method table, or by comparing two method tables, and if not, what sort of optimized comparison can be done, if any.

> We do not need to call `compareTypesForEquality`. The only case where it is not `hnd1 == hnd2` is for type handles with shared generics which we cannot see here (they should be runtime lookups).

It is needed during `gtFoldTypeCompare` because we can fetch a compile time handle even in runtime lookup cases via `gtGetHelperArgClassHandle`. We toss this info away during morph. Wonder if we should keep it around longer.

Reply:
> Yes, it is a little confusing. There are 4 similar concepts in play
> 
> * runtime type
> * compile time handle
> * embedded handle
> * method table
> 
> `compareTypesForEquality` is asking about whether one can infer (at compile time) equivalence of runtime types from compile time handles.

This seems to contradict the naming/type signatures of these functions. If we take `TYPEHANDLE_TO_RUNTIMETYPE` on its name then a "runtime type" is a managed `System.RuntimeType` object. There is also e.g. `CLASSID_RUNTIME_TYPE` that refers specifically to this class. On the other hand, what `compareTypesForEquality` compares are type handles which may or may not require runtime comparison due to being instantiated with `__Canon`.

With this view, `compareTypesForEquality` has nothing to do with the managed objects returned by `TYPEHANDLE_TO_RUNTIMETYPE`.

> It is needed during gtFoldTypeCompare because we can fetch a compile time handle even in runtime lookup cases via gtGetHelperArgClassHandle. We toss this info away during morph. Wonder if we should keep it around longer.

Another source of confusion here is that we have `GenTreeIntCon::gtCompileTimeHandle`, which seems to just be the representation used for communication on the JIT-EE interface. That is a separate concept to e.g. the shared-generics type handles we may get from runtime lookup nodes, which I guess is what you are referring to. I do see that in some cases, with help from the EE `gtFoldTypeCompare` is able to fold type comparisons involving shared generics when the types are "disjoint".

Reply:
I'm afraid you've lost me here. 

`compareTypesForEquality` was added (see https://github.com/dotnet/coreclr/pull/14317) to deduce when comparison of compile time handles would give the same result as runtime code that would invoke the helper on the associated embedded handles / runtime lookups and then compare the resulting runtime type objects.

Reply:
Ok, the way the JIT uses the function makes more sense if `compareTypesForEquality` is meant to compare the runtime type objects for two type handles. Thanks.